protocol/message_persistence.go

```diff
@@ -1331,6 +1449,228 @@ func (db sqlitePersistence) MessageByChatIDs(chatIDs []string, currCursor string
 	return result, newCursor, nil
 }
 
+func (db sqlitePersistence) AllMessagesForBackup() ([]*protobuf.BackedUpMessage, error) {
+	where := "WHERE NOT(m1.hide)"
+	fields := db.tableUserMessagesProtobufFields()
+	selectQuery := `SELECT    %s
+				FROM      user_messages m1
+				LEFT JOIN pin_messages pm
+				ON 	      m1.id = pm.message_id AND pm.pinned = 1`
+	query := fmt.Sprintf(selectQuery, fields) + " " + where
+	rows, err := db.db.Query(query)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	var messages []*protobuf.BackedUpMessage
+	for rows.Next() {
+		message := &protobuf.BackedUpMessage{}
+		if err := db.tableUserMessagesScanProtobufFields(rows, message); err != nil {
+			return nil, err
+		}
+
+		messages = append(messages, message)
+	}
+
+	return messages, nil
+}
+
+func (db sqlitePersistence) saveBackedUpMessages(messages []*protobuf.BackedUpMessage) error {
+	if len(messages) == 0 {
+		return nil
+	}
+
+	tx, err := db.db.BeginTx(context.Background(), &sql.TxOptions{})
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err == nil {
+			err = tx.Commit()
+			return
+		}
+		// don't shadow original error
+		_ = tx.Rollback()
+	}()
+
+	allFields := db.tableUserMessagesAllFields()
+	valuesVector := strings.Repeat("?, ", db.tableUserMessagesAllFieldsCount()-1) + "?"
+	query := "INSERT OR REPLACE INTO user_messages(" + allFields + ") VALUES (" + valuesVector + ")" //nolint: gosec
+	stmt, err := tx.Prepare(query)
+	if err != nil {
+		return err
+	}
+	defer stmt.Close()
+
+	for _, msg := range messages {
+		allValues, err := db.backedUpMessageToUserMessageValues(msg)
+		if err != nil {
+			return err
+		}
+
+		_, err = stmt.Exec(allValues...)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (db sqlitePersistence) saveBackedUpPinMessages(messages []*protobuf.BackedUpMessage) error {
+	pinMessages := make([]*common.PinMessage, 0)
+
+	for _, msg := range messages {
+		if msg.PinnedBy != "" {
+			pinMessage := protobuf.PinMessage{
+				Clock:       msg.Timestamp,
+				MessageId:   msg.Id,
+				ChatId:      msg.ChatId,
+				MessageType: msg.MessageType,
+				Pinned:      true,
+			}
+			pinMessages = append(pinMessages, &common.PinMessage{
+				ID:               msg.Id, // TODO do we need a special ID?
+				From:             msg.PinnedBy,
+				PinMessage:       &pinMessage,
+				WhisperTimestamp: msg.Timestamp,
+			})
+		}
+	}
+	if len(pinMessages) > 0 {
+		err := db.SavePinMessages(pinMessages)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
```

Comment:
I guess you're not a fan of early return pattern 😄 
```suggestion
	for _, msg := range messages {
		if msg.PinnedBy == "" {
      		continue
      	}
		pinMessage := protobuf.PinMessage{
			Clock:       msg.Timestamp,
			MessageId:   msg.Id,
			ChatId:      msg.ChatId,
			MessageType: msg.MessageType,
			Pinned:      true,
		}
		pinMessages = append(pinMessages, &common.PinMessage{
			ID:               msg.Id, // TODO do we need a special ID?
			From:             msg.PinnedBy,
			PinMessage:       &pinMessage,
			WhisperTimestamp: msg.Timestamp,
		})
	}

    if len(pinMessages) == 0 {
        return nil
    }

    return db.SavePinMessages(pinMessages)
```

Reply:
I usually love early returns. If you check my old PR reviews, especially on Embark, it was all "reverse condition and early return please". I'm not sure what happened, maybe I just got old :joy: